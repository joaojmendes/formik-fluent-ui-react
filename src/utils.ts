import { FieldProps, getIn } from 'formik'

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export function getErrorMessage<T = any>({ field, form }: FieldProps<T>) {
  const error = getIn(form.errors, field.name)
  const touched = getIn(form.touched, field.name)

  return touched ? error : undefined
}

export function invokeAll(...callbacks: any[]) {
  return () => {
    for (const callback of callbacks) {
      if (callback && typeof callback === 'function') {
        callback()
      }
    }
  }
}

export function createFakeEvent({ name }: { name: string }) {
  return { target: { name } }
}
