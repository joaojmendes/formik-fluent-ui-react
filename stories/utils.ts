import { action } from '@storybook/addon-actions'
import { FormikActions } from 'formik'

export function handleSubmit<T>(
  values: T,
  { setSubmitting }: FormikActions<T>
) {
  action('submit:start')(values)
  setTimeout(() => {
    setSubmitting(false)
    action('submit:done')()
  }, 2000)
}
