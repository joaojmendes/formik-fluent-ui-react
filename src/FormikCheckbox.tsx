import { FieldProps } from 'formik'
import { Checkbox, ICheckboxProps } from '@fluentui/react'
import * as React from 'react'
import { createFakeEvent, Omit } from './utils'

export function mapFieldToCheckbox<T = any>({
  form,
  field: { value, onChange, onBlur, ...field },
}: FieldProps<T>): Pick<ICheckboxProps, 'checked' | 'name' | 'onChange'> {
  return {
    ...field,
    onChange: (_, checked) => {
      form.setFieldValue(field.name, checked)
      onBlur(createFakeEvent(field))
    },
    checked: value,
  }
}

export type FormikCheckboxProps<T = any> = Omit<
  ICheckboxProps,
  'value' | 'checked' | 'name' | 'onChange' | 'onBlur' | 'form'
> &
  FieldProps<T>

export function FormikCheckbox<T = any>({
  field,
  form,
  ...props
}: FormikCheckboxProps<T>) {
  return <Checkbox {...props} {...mapFieldToCheckbox({ field, form })} />
}
