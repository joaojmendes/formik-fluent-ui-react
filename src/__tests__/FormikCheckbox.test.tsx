import { Field, FieldProps, Form, Formik } from 'formik'
import { Checkbox, setIconOptions } from '@fluentui/react'
import * as React from 'react'
import renderer from 'react-test-renderer'
import { FormikCheckbox, mapFieldToCheckbox } from '../FormikCheckbox'
import { noop, serialize } from './utils'

// Suppress icon warnings.
setIconOptions({
  disableWarnings: true,
})

class Values {
  public isChecked: boolean = false
}

function createFieldProps(value: boolean = false): FieldProps<boolean> {
  return {
    field: {
      value,
      onChange: jest.fn(),
      onBlur: jest.fn(),
      name: 'isChecked',
    },
    form: { setFieldValue: jest.fn(), handleBlur: jest.fn(() => jest.fn()) },
  } as any
}

test('<FormikCheckbox /> renders correctly as a field component', () => {
  const component = renderer.create(
    <Formik initialValues={new Values()} onSubmit={noop}>
      <Form>
        <Field name="isChecked" label="Checkbox" component={FormikCheckbox} />
      </Form>
    </Formik>
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('<FormikCheckbox /> renders a Fabric <Checkbox />', () => {
  const fieldProps = createFieldProps()

  const formikCheckbox = renderer.create(
    <FormikCheckbox {...fieldProps} label="Checkbox" />
  )
  const fabricCheckbox = renderer.create(
    <Checkbox {...mapFieldToCheckbox(fieldProps)} label="Checkbox" />
  )
  expect(serialize(formikCheckbox)).toBe(serialize(fabricCheckbox))
})

test('mapFieldToCheckbox() maps FieldProps to ICheckboxProps', () => {
  const { field, form, meta } = createFieldProps()
  const props = mapFieldToCheckbox({ form, field, meta })

  expect(props.checked).toBe(field.value)

  props.onChange!(null as any, true)

  expect(form.setFieldValue).toHaveBeenCalledTimes(1)
  expect(form.setFieldValue).toHaveBeenCalledWith(field.name, true)

  // onChange also calls onBlur
  expect(field.onBlur).toHaveBeenCalled()
})
