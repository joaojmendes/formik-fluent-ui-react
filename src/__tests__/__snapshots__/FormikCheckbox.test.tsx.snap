// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormikCheckbox /> renders correctly as a field component 1`] = `
<form
  onReset={[Function]}
  onSubmit={[Function]}
>
  <button
    aria-checked={true}
    checked={true}
    className="ms-Checkbox is-checked is-enabled root-47"
    id="checkbox-1"
    name="isChecked"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    role="checkbox"
    type="button"
  >
    <label
      className="ms-Checkbox-label label-48"
      htmlFor="checkbox-1"
    >
      <div
        className="ms-Checkbox-checkbox checkbox-49"
      >
        <i
          aria-hidden={true}
          className="ms-Checkbox-checkmark checkmark-52"
          data-icon-name="CheckMark"
          role="presentation"
        />
      </div>
      <span
        className="ms-Checkbox-text text-51"
      >
        Checkbox
      </span>
    </label>
  </button>
</form>
`;
